### PR TITLE
BDRSPS-913 Added site visit id cross validation input to site visit template. 

### PR DESCRIPTION
**Pull request by joecrowleygaia**
Removed `LogicalOr` foreign key logic and added mutual inclusion whitelist logic.

[EDIT] Removed whitelist logic from mutual inclusion and had it inserted into newly created enhanced required custom checker.

### CODE REVIEW

File: tests/templates/test_survey_site_visit_data_v2.py [hunk 1]
```diff
@@ -112,13 +126,19 @@ def mocked_schema(self, mocker: pytest_mock.MockerFixture) -> unittest.mock.Magi
         descriptor["fields"] = [f for f in descriptor["fields"] if f["name"] in fieldnames]
 
         # Patch and return
-        return mocker.patch.object(Mapper, "schema", return_value=descriptor)
+        yield mocker.patch.object(mapping.SurveySiteVisitMapper, "schema", return_value=descriptor)
+
+        # Clear schema lru_cache
+        mapping.SurveySiteVisitMapper.schema.cache_clear()
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply tests/templates/test_survey_site_visit_data_v2.py:132]: ok? I don't mind extra cache_clears here. I had a few headaches with calls to schema not being returned back to normal in separate tests after patching.

[the remark this answers is not in the document]
Comment (Lincoln-GR) [reply tests/templates/test_survey_site_visit_data_v2.py:132]: I meant `.schema` is a mock so `.schema.cache_clear()` will also be a mock and not actually do anything. that was the point I twas trying to make.

File: abis_mapping/templates/survey_site_visit_data_v2/mapping.py [hunk 2]
```diff
@@ -43,34 +43,50 @@ def apply_validation(
...
+            plugins.chronological.ChronologicalOrder(
+                field_names=["siteVisitStart", "siteVisitEnd"],
+            ),
+            plugins.logical_or.LogicalOr(
+                field_names=["siteVisitStart", "siteVisitEnd"],
+            ),
+        ]
+
+        # Check to see if site visit id map was provided
+        if site_visit_id_map:
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/templates/survey_site_visit_data_v2/mapping.py:72]: that's intentional based on default and the case someone supplies an empty dict (treating them the same).

File: tests/templates/test_survey_site_visit_data_v2.py [hunk 3]
```diff
@@ -112,13 +126,19 @@ def mocked_schema(self, mocker: pytest_mock.MockerFixture) -> unittest.mock.Magi
         descriptor["fields"] = [f for f in descriptor["fields"] if f["name"] in fieldnames]
```
Comment (Lincoln-GR) [inline tests/templates/test_survey_site_visit_data_v2.py:126]: This line is probably the source of schema not going back to normal.
It is mutating the `descriptor` dict which was returned by `schema()` and is stored in the cache.

File: abis_mapping/templates/survey_site_visit_data_v2/mapping.py [hunk 4]
```diff
@@ -43,34 +43,50 @@ def apply_validation(
             frictionless.Report: Validation report for the specified data.
         """
         # Extract keyword arguments
-        # TODO: Uncomment
-        # site_visit_id_map: dict[str, bool] = kwargs.get("site_visit_id_map", {})
+        site_visit_id_map: dict[str, bool] = kwargs.get("site_visit_id_map", {})
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/templates/survey_site_visit_data_v2/mapping.py:46]: i think i put those into the same basket, hence the empty dict default.

[the remark this answers is not in the document]
Comment (Lincoln-GR) [reply abis_mapping/templates/survey_site_visit_data_v2/mapping.py:46]: Should we treat them differently?
not provided (`None`) = site visit file being validated on its own, vs
empty dict = no site visit ids present in the occurrence file

File: abis_mapping/plugins/mutual_inclusion.py [hunk 5]
```diff
@@ -40,11 +46,26 @@ def validate_row(self, row: frictionless.Row) -> Iterator[frictionless.Error]:
...
+        # If there is at least one item in the dictionary then it is deemed valid.
+        if len(row_fk_map) > 0:
+            return
+
+        note = f"the columns {self.field_names} are mutually inclusive and values must be provided together "
+
+        if len(self.whitelists) > 0:
+            note += (
+                f" or a given value must be referenced by one of the supplied whitelist fields"
+                f" {self.whitelists.keys()}"
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/plugins/mutual_inclusion.py:63]: Any suggestions?

File: abis_mapping/templates/survey_site_visit_data_v2/mapping.py [hunk 6]
```diff
@@ -43,34 +43,50 @@ def apply_validation(
...
+            ),
+            plugins.logical_or.LogicalOr(
+                field_names=["siteVisitStart", "siteVisitEnd"],
+            ),
+        ]
+
+        # Check to see if site visit id map was provided
+        if site_visit_id_map:
+            # Construct foreign key map
+            fk_map = {"siteVisitID": set(site_visit_id_map)} if site_visit_id_map else dict()
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/templates/survey_site_visit_data_v2/mapping.py]: yep. removed.

File: abis_mapping/templates/survey_site_visit_data_v2/mapping.py [hunk 7]
```diff
@@ -43,34 +43,50 @@ def apply_validation(
...
+
+            # Add custom check for temporal flexibility with whitelists. Here deferring
+            # the check on any ids found in the occurrence template, to when validaation
+            # occurs on it in line with temporal flexibility rules.
+            checks += [
+                plugins.mutual_inclusion.MutuallyInclusive(
+                    field_names=["siteVisitID", "siteVisitStart"],
+                    whitelists=fk_map,
+                )
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/templates/survey_site_visit_data_v2/mapping.py:83]: I think my preference might be for the other way, and get rid of the default lookup custom check.

File: tests/templates/test_survey_site_visit_data_v2.py (hunk 3 above)
Comment (joecrowleygaia) [reply tests/templates/test_survey_site_visit_data_v2.py:126]: i thought that too and attempted to use `copy.deepcopy` to isolate it. Didn't work.

File: abis_mapping/templates/survey_site_visit_data_v2/schema.json [hunk 8]
```diff
@@ -1,165 +1,174 @@
...
+        }
+    ],
+    "foreignKeys": [
+        {
+            "fields": "siteVisitID",
+            "reference": {
+                "resource": "survey_occurrence_data",
+                "fields": "siteVisitID"
+            }
+        }
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/templates/survey_site_visit_data_v2/schema.json]: I was going to use the frictionless builtin foreign key functionality against a newly created resource called "survey_occurrence_data" that would be made from the supplied id map but in the end wasn't the way to go. The reference "survey_occurrence_data" doesn't really relate to anything. They don't do anything currently. 

Removed.

File: abis_mapping/templates/survey_site_visit_data_v2/mapping.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/templates/survey_site_visit_data_v2/mapping.py:46]: no

File: tests/templates/test_survey_site_visit_data_v2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply tests/templates/test_survey_site_visit_data_v2.py]: yeah you're right. Removed.

File: abis_mapping/plugins/mutual_inclusion.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/plugins/mutual_inclusion.py]: removed from check and  put into `RequiredEnhanced` check.

File: abis_mapping/templates/survey_site_visit_data_v2/mapping.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/templates/survey_site_visit_data_v2/mapping.py]: created new custom check `RequiredEnhanced` which achieves the same thing, with slightly better error wording and no need to include "siteVisitID" in `field_names` arg.

**Review by Lincoln-GR — COMMENTED** [review]
Looks good, two smalll thins

File: abis_mapping/plugins/required.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        if len(bypassable_values) > 0:
+            return
+
+        note = f"the columns {self.field_names} are all required"
+        if self.whitelists:
+            note += (
+                f" unless the values provided in fields {self.whitelists.keys()}"
+                " match any of those in their supplied whitelists"
+            )
+        note += f"({missing} are missing)."
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/plugins/required.py]: fixed

File: abis_mapping/plugins/mutual_inclusion.py [hunk 10]
```diff
@@ -40,11 +40,13 @@ def validate_row(self, row: frictionless.Row) -> Iterator[frictionless.Error]:
             # Short-circuit
             return
 
+        note = (
+            f"the columns {self.field_names} are mutually inclusive and values"
+            " must be provided together (columns {missing} are missing values)"
```
[the remark this answers is not in the document]
Comment (joecrowleygaia) [reply abis_mapping/plugins/mutual_inclusion.py]: fixed

**Review by Lincoln-GR — APPROVED** [review]

**Pull request merged**